### PR TITLE
os/bluestore: Hint based allocation in bitmap Allocator

### PR DESCRIPTION
**Pull request by chhabaramesh**
This pull request has following changes:
1. Hint based allocation in BitMapAllocator.
2. Unit Test cases related to that.
3. Cleanup of wrap argument that is no more in use.

### CODE REVIEW

**Comment by chhabaramesh** [comment]
@liewegas , Please review this.

File: src/test/objectstore/Allocator_test.cc [hunk 1]
```diff
@@ -187,6 +187,49 @@ TEST_P(AllocTest, test_alloc_failure)
   }
 }
 
+TEST_P(AllocTest, test_alloc_hint_bmap)
+{
+  if (GetParam() == std::string("stupid")) {
```
Comment (liewegas) [inline src/test/objectstore/Allocator_test.cc]: != std::string("bitmap") ?

**Comment by liewegas** [comment]
@chhabaramesh I haven't been following bitmap allocator internals.. is there someone else that can review?

**Comment by chhabaramesh** [comment]
@liewegas , I can explain you it to you in short call ? Or Maybe @ifed01 or Manavalan can review?

But this change is quite simple, just makes sure that search of bit start from a given hint.

File: src/test/objectstore/Allocator_test.cc (hunk 1 above)
Comment (chhabaramesh) [inline src/test/objectstore/Allocator_test.cc]: The stupid allocator was not passing all new hint test cases. So added this to bypass it for stupid allocator.

**Comment by liewegas** [comment]
I see where the hint is used, but I don't see where the search loops around to the beginning if it doesn't find anything > hint (and, ideally, only searches back up to the start point).

**Comment by chhabaramesh** [comment]
@liewegas The iterator BmapEntityListIter takes a wrap argument and that is true only for children of root nodes.

1555 int64_t BitAllocator::alloc_blocks_dis_int(bool wait, int64_t num_blocks,
1556            int64_t hint, int64_t area_blk_off, ExtentList *block_list)
1557 {
1558   return alloc_blocks_dis_int_work(wait, true, num_blocks, hint,
1559                      area_blk_off, block_list);
1560 }

At all other levels it should be false. So it will start at a child at root level and wrap around and end at same child.

I think there is a type bug at:

 959 int64_t BitMapAreaIN::alloc_blocks_dis_int(bool wait, int64_t num_blocks,
 960            int64_t hint, int64_t area_blk_off, ExtentList *block_list)
 961 {
 962   return alloc_blocks_dis_int_work(wait, true, num_blocks, hint,
 963                      area_blk_off, block_list);
 964 }

The wrap ( second ) argument must be false here :(.

**Comment by chhabaramesh** [comment]
@liewegas , Fixed the typo bug mentioned above and updated. Please check again.

**Pull request merged**